src/workers/python/papyros/papyros.py

```diff
@@ -112,41 +98,25 @@ def pre_run(self, source_code, mode="exec", top_level_await=False):
         self.override_globals()
         return super().pre_run(source_code, mode=mode, top_level_await=top_level_await)
 
-    def execute(self, code_obj, source_code, mode=None):  # noqa
-        return eval(code_obj, self.console.locals)  # noqa
-
-    async def run_async(self, source_code, mode="exec", top_level_await=True, filename="my_code.py"):
-        """
-        Mostly a copy of the parent `run_async` with some key differences
-        We use `await ft` in case of an exception, because `serialize_traceback` isn't async.
-        We call pre_run within to handle its exceptions within this try-block too
-        As it will rethrow the SyntaxError (@see serialize_syntax_error)
-        """
-        self.set_file(filename, source_code)
-        try:
-            with self._execute_context(source_code):
-                try:
-                    await self.install_imports(source_code, ignore_missing=False)
-                    code_obj = self.pre_run(
-                        source_code, mode, top_level_await=top_level_await)
-                    if code_obj:
-                        result = self.execute(code_obj, source_code, mode)
-                        while isinstance(result, Awaitable):
-                            result = await result
-                        return result
-                except:
-                    # Let `_execute_context` and `serialize_traceback`
-                    # handle the exception
-                    raise
-        except KeyboardInterrupt:
-            self.callback("interrupt", data="KeyboardInterrupt", contentType="text/plain")
-
-    def serialize_syntax_error(self, exc, source_code):
+    async def run_async(self, source_code, mode="exec", top_level_await=True):
+        with self._execute_context():
+            try:
+                await self.install_imports(source_code, ignore_missing=False)
+                code_obj = self.pre_run(source_code, mode=mode, top_level_await=top_level_await)
+                if code_obj:
+                    result = self.execute(code_obj, mode)
+                    while isinstance(result, Awaitable):
+                        result = await result
+                    return result
+            except KeyboardInterrupt:
+                self.callback("interrupt", data="KeyBoardInterrupt", contentType="text/plain")
```

Comment:
Are you actually using this? At this point you already know the program is ending. Either way, if you handle this in serialize_traceback then you should be able to avoid copying run_async.

Reply:
The KeyboardInterrupt is captured in _execute_context via
```python
except BaseException as e:
    self.output("traceback", **self.serialize_traceback(e))
```
I could also check in the event translation part for a KeyboardInterrupt error name, but output expects strings (which is why my FriendlyErrors are still dumped to strings using json) so the check would not really be clean (would need json parsing to cover all edge cases). If you have a possible fix for this, then I will gladly incorporate it.

Reply:
I meant you could just do this in `serialize_traceback`:

```python
if isinstance(e, KeyboardInterrupt):
    self.callback("interrupt", ...)
    # maybe return reduced output here
```

assuming that event type is valuable somehow.

> output expects strings (which is why my FriendlyErrors are still dumped to strings using json)

You could set text to `""` or a simple formatted traceback, then return the actual data in a dict without dumping to json.

The output system could be a bit less opinionated and not force you to supply `text`. `OutputBuffer` concatenates `text` values where possible, but realistically that only happens with stdout/stderr and maybe custom output types, not `traceback`. futurecoder concatenates the text from all parts into a single string which is useful for a few things, so it's nice when `text` always contains something human readable, but that's futurecoder.